File: apps/Card/android/core/src/main/java/com/americangroupllc/card/core/domain/CardKindTransitions.kt
package com.americangroupllc.card.core.domain

import com.americangroupllc.card.core.models.Card
import com.americangroupllc.card.core.models.CardKind

/**
 * Pure logic for note ↔ task ↔ reminder transitions. Mirrors
 * shared/CardCore/Sources/CardCore/Domain/CardKindTransitions.swift
 */
object CardKindTransitions {

    /**
     * Convert [card] to [target]. Clears the irrelevant fields per kind.
     * For [CardKind.REMINDER], [reminderAtEpochMs] MUST be in the future or
     * this returns null.
     */
    fun convert(
        card: Card,
        target: CardKind,
        reminderAtEpochMs: Long? = null,
        nowEpochMs: Long = System.currentTimeMillis(),
    ): Card? {
        if (card.kind == target && target != CardKind.REMINDER) {
            return card
        }
        return when (target) {
            CardKind.NOTE -> card.copy(
                kind = CardKind.NOTE,
                reminderAtEpochMs = null,
                completedAtEpochMs = null,
                updatedAtEpochMs = nowEpochMs,
            )
            CardKind.TASK -> card.copy(
                kind = CardKind.TASK,
                reminderAtEpochMs = null,
                updatedAtEpochMs = nowEpochMs,
            )
            CardKind.REMINDER -> {
                if (reminderAtEpochMs == null || reminderAtEpochMs <= nowEpochMs) return null
                card.copy(
                    kind = CardKind.REMINDER,
                    reminderAtEpochMs = reminderAtEpochMs,
                    completedAtEpochMs = null,
                    updatedAtEpochMs = nowEpochMs,
                )
            }
        }
    }

    /** Toggle a task's completion state. No-op for non-tasks. */
    fun toggleCompleted(
        card: Card,
        nowEpochMs: Long = System.currentTimeMillis(),
    ): Card {
        if (card.kind != CardKind.TASK) return card
        return card.copy(
            completedAtEpochMs = if (card.isCompleted) null else nowEpochMs,
            updatedAtEpochMs = nowEpochMs,
        )
    }
}

File: apps/Card/shared/CardCore/Sources/CardCore/Domain/CardKindTransitions.swift
// CardKindTransitions — pure logic for note ↔ task ↔ reminder transitions.
// Mirrored in android/core/.../domain/CardKindTransitions.kt
import Foundation

public enum CardKindTransitions {
    /// Convert `card` to `target`. Clears the irrelevant fields per kind.
    /// For `.reminder`, `reminderAt` MUST be in the future or this returns `nil`.
    public static func convert(
        _ card: Card,
        to target: CardKind,
        reminderAt: Date? = nil,
        now: Date = Date()
    ) -> Card? {
        if card.kind == target && target != .reminder {
            // Identity transition for non-reminder kinds is a no-op.
            return card
        }

        var updated = card
        updated.kind = target
        updated.updatedAt = now

        switch target {
        case .note:
            updated.reminderAt = nil
            updated.completedAt = nil
        case .task:
            updated.reminderAt = nil
            // completedAt preserved if already set
        case .reminder:
            guard let when = reminderAt, when > now else { return nil }
            updated.reminderAt = when
            updated.completedAt = nil
        }
        return updated
    }

    /// Toggle a task's completion state. No-op for non-tasks.
    public static func toggleCompleted(_ card: Card, now: Date = Date()) -> Card {
        guard card.kind == .task else { return card }
        var updated = card
        updated.completedAt = card.isCompleted ? nil : now
        updated.updatedAt = now
        return updated
    }
}
